base/usr/share/man/man1/weather-tool.1
.TH WEATHER-TOOL 1 "April, 2026"

.SH NAME
\fBweather\-tool\fP \- update cached weather forecast data

.SH SYNOPSIS
\fBweather\-tool\fP

.SH DESCRIPTON
This utility fetches weather forecast data from OpenWeatherMap using a stored configuration,
parses the JSON data from that API, and creates a plain-text file at \fB/tmp/weather\-parsed.conf\fP
which can then be read by other utilities such as the panel weather widget.

.SH FILES
The utility reads configuration data from \fB/etc/weather.json\fP.
.PP
It will attempt to read location data from \fB/tmp/location\-data.json\fP, or query
an API provided by \fBip\-api.com\fP to supply that data. Data queried from this
third\-party service includes the latitude, longitude, city name, and timezone
offset associated with the user's IP address in the third\-party's geolocation database.
.PP
The utility stores the raw results of weather data queries at \fB/tmp/weather\-data.json\fP.
Tools that wish to read the JSON data directly may obtain it from that location.
.PP
The utility stores its final parsed results in \fB/tmp/weather\-parsed.conf\fP.
.PP
This file is formatted with line\-separated data as follows:

    \(bu The formatted temperature, with two decimal points.
    \(bu The temperature as an integer.
    \(bu A string describing the weather conditions, such as "Clouds"
    \(bu An icon identifier for the weather conditins. See upstream docs for more info.
    \(bu The humidity as integer percentage.
    \(bu The cloud coverage as an integer percentage.
    \(bu The city name acquired from \fBip\-api.com\fP.
    \(bu The date of the last update.

.SH THIRD PARTY APIS
This utility communicates with two third\-party remote APIs over the Internet:

.TP
.B OpenWeatherMap
Provides weather data. This API requires a key, which is provided with the OS.

.TP
.B ip\-api.com
Provides location and timezone data.

.PP
The information obtained from these APIs is not shared with the ToaruOS maintainers
at any point. Users should review the respective services' privacy policies and terms
before exposing ToaruOS to an Internet connection, and may blackhole or provide
alternative implementations of these APIs if they desire.

.PP
The availability of these APIs is subject to change.

.SH SEE ALSO
.BR weather\-configurator (1),
.BR find\-timezone (1),
.BR panel (1)


apps/weather-tool.c
/**
 * @brief Ask OpenWeather for forecast data.
 *
 * Fetches weather forecast data from OpenWeather and converts the JSON
 * output to a simpler parsed format we read in the panel.
 *
 * @copyright
 * This file is part of ToaruOS and is released under the terms
 * of the NCSA / University of Illinois License - see LICENSE.md
 * Copyright (C) 2018-2021 K. Lange
 */
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>
#include <unistd.h>
#include <sys/time.h>
#include <toaru/json.h>
#include <toaru/hashmap.h>
#include <toaru/list.h>

typedef struct JSON_Value Value;

#define WEATHER_CONF_PATH  "/etc/weather.json"
#define WEATHER_DATA_PATH  "/tmp/weather-data.json"
#define WEATHER_OUT_PATH   "/tmp/weather-parsed.conf"
#define LOCATION_DATA_PATH "/tmp/location-data.json"

int main(int argc, char * argv[]) {
	Value * config = json_parse_file(WEATHER_CONF_PATH);
	if (!config) {
		fprintf(stderr, "No weather config data\n");
		return 1;
	}

	char * city = JSON_KEY(config, "city")->string;
	char * key = JSON_KEY(config, "key")->string;
	char * units = JSON_KEY(config, "units")->string;
	char cmdline[1024];
	
	/* Check if the unit input is invalid (see #222)
	 * 	This should in fact be checked before the configuration is saved
	 *	and inform the user of an invalid value */
	if (((strcmp(units, "metric") || (strcmp(units, "imperial")) {
		units = "metric";
	}

	/* If the city is 'guess', we'll make a single query to a separate service to
	 * get a location from the user's external IP... */
	if (!strcmp(city, "guess")) {
		/* See if the location data already exists... */
		Value * locationData = json_parse_file(LOCATION_DATA_PATH);
		if (!locationData) {
			sprintf(cmdline, "fetch -o \"" LOCATION_DATA_PATH "\" \"http://ip-api.com/json/?fields=lat,lon,city,offset\"");
			system(cmdline);
			locationData = json_parse_file(LOCATION_DATA_PATH);
		}
		/* If we still failed to load it, then bail. */
		if (!locationData) {
			fprintf(stderr, "%s: city field was set to 'guess' but failed to acquire data from IP geolocation service\n", argv[0]);
			return 1;
		}

		city = JSON_KEY(locationData, "city")->string;
		double lat = JSON_KEY(locationData, "lat")->number;
		double lon = JSON_KEY(locationData, "lon")->number;

		sprintf(cmdline, "fetch -o \"" WEATHER_DATA_PATH "\" \"http://api.openweathermap.org/data/2.5/weather?lat=%.5lf&lon=%.5lf&appid=%s&units=%s\"", lat, lon, key, units);
		system(cmdline);
	} else {
		sprintf(cmdline, "fetch -o \"" WEATHER_DATA_PATH "\" \"http://api.openweathermap.org/data/2.5/weather?q=%s&appid=%s&units=%s\"", city, key, units);
		system(cmdline);
	}

	Value * result = json_parse_file(WEATHER_DATA_PATH);
	assert(result && result->type == JSON_TYPE_OBJECT);

	Value * _main = JSON_KEY(result,"main");
	Value * conditions = (JSON_KEY(result,"weather") && JSON_KEY(result,"weather")->array->length > 0) ?
		JSON_IND(JSON_KEY(result,"weather"),0) : NULL;

	FILE * out = fopen(WEATHER_OUT_PATH, "w");

	/**
	 * The format for a parsed weather payload is a series of line-separated entries:
	 * - Formatted temperature, with decimal.
	 * - Integral temperature, eg. for the panel widget.
	 * - Main weather conditions string, eg. "Clouds"
	 * - Icon identifier, eg. 02d is "cloudy, daytime".
	 * - Humidity (integer, percentage)
	 * - Cloud coverage (integer, percentage)
	 * - City name (we're using the guessed location, not the one from the weather provider...)
	 * - Date string of last update
	 */
	fprintf(out, "%.2lf\n", JSON_KEY(_main,"temp")->number);
	fprintf(out, "%d\n", (int)JSON_KEY(_main,"temp")->number);
	fprintf(out, "%s\n", conditions ? JSON_KEY(conditions,"main")->string : "");
	fprintf(out, "%s\n", conditions ? JSON_KEY(conditions,"icon")->string : "");
	fprintf(out, "%d\n", (int)JSON_KEY(_main,"humidity")->number);
	fprintf(out, "%d\n", JSON_KEY(JSON_KEY(result,"clouds"),"all") ? (int)JSON_KEY(JSON_KEY(result,"clouds"),"all")->number : 0);
	fprintf(out, "%s\n", city);
	char * format = "%a, %d %b %Y %H:%M:%S\n";
	struct tm * timeinfo;
	struct timeval now;
	char buf[BUFSIZ] = {0};
	gettimeofday(&now, NULL);
	timeinfo = localtime((time_t *)&now.tv_sec);
	strftime(buf,BUFSIZ,format,timeinfo);
	fprintf(out, buf);

	fprintf(out, "%d\n", (int)JSON_KEY(_main,"pressure")->number);

	fclose(out);

	return 0;
}

